Release/src/http/client/http_client_named_pipe.cpp

```diff
@@ -0,0 +1,470 @@
+enum msg_body_type
+{
+    no_body,
+    content_length_chunked,
+    transfer_encoding_chunked
+};
+
+// Additional information necessary to track a WinHTTP request.
+class named_pipe_request_context : public request_context
+{
+public:
+
+    // Factory function to create requests on the heap.
+    static std::shared_ptr<request_context> create_request_context(const std::shared_ptr<_http_client_communicator> &client, const http_request &request)
+    {
+        std::shared_ptr<named_pipe_request_context> ret(new named_pipe_request_context(client, request));
+        ret->m_self_reference = ret;
+        return std::move(ret);
+    }
+
+    ~named_pipe_request_context()
+    {
+        cleanup();
+    }
+
+    msg_body_type m_bodyType;
+
+    utility::size64_t m_remaining_to_write;
+
+    std::char_traits<uint8_t>::pos_type m_startingPosition;
+
+    // If the user specified that to guarantee data buffering of request data, in case of challenged authentication requests, etc...
+    // Then if the request stream buffer doesn't support seeking we need to copy the body chunks as it is sent.
+    concurrency::streams::istream m_readStream;
+    std::unique_ptr<concurrency::streams::container_buffer<std::vector<uint8_t>>> m_readBufferCopy;
+    virtual concurrency::streams::streambuf<uint8_t> _get_readbuffer()
+    {
+        return m_readStream.streambuf();
+    }
+
+    // This self reference will keep us alive until finish() is called.
+    std::shared_ptr<named_pipe_request_context> m_self_reference;
+    memory_holder m_body_data;
+    std::unique_ptr<web::http::details::compression::stream_decompressor> decompressor;
+
+    virtual void cleanup()
+    {
+        // TODO
+    }
+
+protected:
+
+    virtual void finish()
+    {
+        request_context::finish();
+        assert(m_self_reference != nullptr);
+        auto dereference_self = std::move(m_self_reference);
+        // As the stack frame cleans up, this will be deleted if no other references exist.
+    }
+
+private:
+
+    // Can only create on the heap using factory function.
+    named_pipe_request_context(const std::shared_ptr<_http_client_communicator> &client, const http_request &request)
+        : request_context(client, request),
+        m_bodyType(no_body),
+        m_startingPosition(std::char_traits<uint8_t>::eof()),
+        m_body_data(),
+        m_remaining_to_write(0),
+        m_readStream(request.body())
+    {
+    }
+};
+
+// named pipe client.
+class named_pipe_client : public _http_client_communicator
+{
+public:
+    static const DWORD PIPE_BUFFER_SIZE = 64 * 1024;
+
+    named_pipe_client(http::uri address, http_client_config client_config)
+        : _http_client_communicator(std::move(address), std::move(client_config))
+    {}
+
+    named_pipe_client(const named_pipe_client&) = delete;
+    named_pipe_client &operator=(const named_pipe_client&) = delete;
+
+    // Closes session.
+    ~named_pipe_client()
+    {
+        // TODO
```

Comment:
the returned async task from the API's below are sharing 'this' object. how are we ensuring the tasks are not trying to access destructed instance.

Reply:
It's a shared pointer - last guy will destruct